varnish-macros/src/stats.rs

```diff
@@ -0,0 +1,162 @@
+    pub format: String,   // "integer", "bytes", "duration", "bitmap", etc
+    pub docs: String,
+    pub index: Option<usize>,
+}
+
+#[derive(Serialize)]
+struct VscMetadata {
+    version: String,
+    name: String,
+    oneliner: String,
+    order: u32,
+    docs: String,
+    elements: usize,
+    elem: std::collections::HashMap<String, VscMetricDef>,
+}
+
+pub fn get_struct_fields(data: &Data) -> &FieldList {
+    match data {
+        Data::Struct(data) => match &data.fields {
+            Fields::Named(fields) => &fields.named,
+            _ => panic!("Only named fields are supported"),
+        },
+        _ => panic!("Only structs are supported"),
+    }
+}
+
+pub fn validate_fields(fields: &FieldList) {
+    for field in fields {
+        match &field.ty {
+            Type::Path(path) => {
+                let is_atomic_u64 = path
+                    .path
+                    .segments
+                    .last()
+                    .is_some_and(|seg| seg.ident == "AtomicU64");
+
+                if !is_atomic_u64 {
+                    let field_name = field.ident.as_ref().unwrap();
+                    panic!("Field {field_name} must be of type AtomicU64");
+                }
+            }
+            _ => panic!("Field types must be AtomicU64"),
+        }
+    }
+}
+
+fn generate_metrics(fields: &FieldList) -> Vec<VscMetricDef> {
+    fields
+        .iter()
+        .enumerate()
+        .map(|(i, field)| {
+            let name = field.ident.as_ref().unwrap().to_string();
+
+            let counter_type = if field
+                .attrs
+                .iter()
+                .any(|attr| attr.path().is_ident("counter"))
+            {
+                "counter"
+            } else if field.attrs.iter().any(|attr| attr.path().is_ident("gauge")) {
+                "gauge"
+            } else {
+                panic!("Field {name} must have either #[counter] or #[gauge] attribute")
+            };
+
+            let doc_str = parse_doc_str(&field.attrs);
+            let mut doc_lines = doc_str.split('\n').filter(|s| !s.is_empty());
+            let oneliner = doc_lines.next().unwrap_or_default().to_string();
+            let docs = doc_lines.next().unwrap_or_default().to_string();
+
+            let (level, format) = parse_counter_attributes(field, counter_type);
+
+            VscMetricDef {
+                name,
+                counter_type: counter_type.to_string(),
+                ctype: "uint64_t".to_string(),
+                level,
+                oneliner,
+                format,
+                docs,
+                index: Some(i * 8),
+            }
+        })
+        .collect()
+}
+
+pub fn generate_metadata_json(name: &str, fields: &FieldList) -> String {
+    let metrics = generate_metrics(fields);
+
+    let metadata = VscMetadata {
+        version: "1".to_string(),
+        name: name.to_string(),
+        oneliner: format!("{name} statistics"),
+        order: 100,
```

Comment:
need to comment the meaning of 100

Reply:
Hrm, I looked into it and it appears to do...nothing? I don't see it getting read in the varnish-cache source. I guess I can just try leaving it off entirely.

Reply:
@gquintard any thoughts on this?